source/extensions/clusters/aggregate/cluster.cc

```diff
@@ -0,0 +1,150 @@
+#include "extensions/clusters/aggregate/cluster.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Clusters {
+namespace Aggregate {
+
+Cluster::Cluster(const envoy::api::v2::Cluster& cluster,
+                 const envoy::config::cluster::aggregate::ClusterConfig& config,
+                 Upstream::ClusterManager& cluster_manager, Runtime::Loader& runtime,
+                 Runtime::RandomGenerator& random,
+                 Server::Configuration::TransportSocketFactoryContext& factory_context,
+                 Stats::ScopePtr&& stats_scope, ThreadLocal::SlotAllocator& tls, bool added_via_api)
+    : Upstream::ClusterImplBase(cluster, runtime, factory_context, std::move(stats_scope),
+                                added_via_api),
+      cluster_manager_(cluster_manager), runtime_(runtime), random_(random),
+      tls_(tls.allocateSlot()) {
+  for (const auto& inner_cluster : config.clusters()) {
+    clusters_.emplace_back(inner_cluster);
+  }
+}
+
+PriorityContext
+Cluster::linearizePrioritySet(const std::function<bool(const std::string&)>& skip_predicate) {
+  Upstream::PrioritySetImpl priority_set;
+  std::vector<std::pair<uint32_t, Upstream::ThreadLocalCluster*>> priority_to_cluster;
+  int next_priority_after_linearizing = 0;
+
+  // Linearize the priority set. e.g. for clusters [C_0, C_1, C_2] referred in aggregate cluster
+  //    C_0 [P_0, P_1, P_2]
+  //    C_1 [P_0, P_1]
+  //    C_2 [P_0, P_1, P_2, P_3]
+  // The linearization result is:
+  //    [C_0.P_0, C_0.P_1, C_0.P_2, C_1.P_0, C_1.P_1, C_2.P_0, C_2.P_1, C_2.P_2, C_2.P_3]
+  // and the traffic will be distributed among these priorities.
+  for (const auto& cluster : clusters_) {
+    if (skip_predicate(cluster)) {
+      continue;
+    }
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
+      continue;
+    }
+
+    int priority_in_current_cluster = 0;
+    for (const auto& host_set : tlc->prioritySet().hostSetsPerPriority()) {
+      if (!host_set->hosts().empty()) {
+        priority_set.updateHosts(
+            next_priority_after_linearizing++, Upstream::HostSetImpl::updateHostsParams(*host_set),
+            host_set->localityWeights(), host_set->hosts(), {}, host_set->overprovisioningFactor());
+        priority_to_cluster.emplace_back(std::make_pair(priority_in_current_cluster, tlc));
+      }
+      priority_in_current_cluster++;
+    }
+  }
+
+  return std::make_pair(std::move(priority_set), std::move(priority_to_cluster));
+}
+
+void Cluster::startPreInit() {
+  for (const auto& cluster : clusters_) {
+    auto tlc = cluster_manager_.get(cluster);
+    if (tlc == nullptr) {
```

Comment:
When does this happen? Can you add a comment?

Reply:
Done